### DISCUSSION TITLE
Why are columns narrow when there is horizontal space and the data is wide?

### DISCUSSION THREAD

**Discussion by frosencrantz** [saulpw/visidata#1486] [General]
Why aren't columns wider when there is horizontal space and the data is wide?   I see this at times with VisiData. When there is empty space on the right,  and there are columns that are narrower than the data in them.

Here is an example from another issue I reported:
   - https://asciinema.org/a/DyJoj2OPxS57yHaKlxeyvWgO7?t=4.6
Here at around 4.6 seconds the columns start out wide, but once the data is loaded, Visidata actually makes the filename column narrower than the filenames.

I bet I could find multiple examples of where a data column's values are wider than the default column width, but there is unused space on the right. 

I was thinking about this in the context of what could be done to make VisiData  "prettier".  And making data visible when there is space, seemed helpful.

**Comment by saulpw** [comment]
> the columns start out wide, but once the data is loaded, Visidata actually makes the filename column narrower than the filenames.

Note that the filename column does not decrease in width.  It is the same width (`options.default_width`) as before the sheet loaded; it's just that other columns shrank to fit their data.  This option is so that columns with very wide data don't become as wide as the screen (which is very annoying).

You might experiment with larger values for `options.default_width`.  Try 50 or 80 or 100 and see if that makes all your use cases look prettier.  I'm open to changing the default value for this option.

**Reply by saulpw** [reply]
Also remember there there may be implied consequences to a column's width; for instance, the width of the fields in when saving to the `fixed` (fixed-width) filetype, is the stated width of the columns.  This is kind of a WYSIWYG design.  So we don't want to change the width once it has already been set.  If we were to do anything, we could make the initial column width calculation (when `Column.width` is None, before it has been set to anything) depend on the full terminal width.  More work though and for perhaps marginal benefit!

**Reply by frosencrantz** [reply]
I think it would be helpful if VisiData could better pick widths of columns.  I think it did a good job for the other columns, but not for the filename column.  I wonder if the column width could be more dynamic somehow, so I wouldn't need to pick a better `options.default_width`.  

> More work though and for perhaps marginal benefit!

I am talking about prettier, which may seem very marginal.  VisiData does do a good job at putting data on the screen.  But there are places like in this case, where it would be useful if there was some dynamic update after the data was loaded (or at least partially loaded).  What is updating the width of the other columns, could that mechanism be used to expand column widths?

Because of the way the fixed-width sheet is parsed, I  wouldn't  expect to see much change in the width of the columns.

**Reply by saulpw** [reply]
> I think it would be helpful if VisiData could better pick widths of columns. I think it did a good job for the other columns, but not for the filename column.

It sounds like you would want the filename column to fit the entire filename.  What if filenames were 128 characters long, should the column be that wide too, or is there some reasonable maximum for this auto-sizing?  This reasonable maximum is what `options.default_width` is exactly about.  If you want the columns to fit everything always, just set it to 1000 or something huge.

> But there are places like in this case, where it would be useful if there was some dynamic update after the data was loaded (or at least partially loaded). What is updating the width of the other columns, could that mechanism be used to expand column widths?

When the `Column.width` is None, on first display it looks at the rows being displayed on the screen.  See https://github.com/saulpw/visidata/blob/develop/visidata/sheets.py#L623.  It then sets the width based on that, so it's not being recomputed all the time (otherwise the column widths may jump around quite erratically when you page down).  So they are only updated once.  We could make that algorithm smarter (to take into account the width of the screen, for example), but making it 'more dynamic' resulted in a lot of terrible behavior that was not worth it (back when this was first implemented).

> Because of the way the fixed-width sheet is parsed, I wouldn't expect to see much change in the width of the columns.

I mean, if you want to save any kind of sheet as fixed-width.  Like if you got some data from a JSON API and want to format it for some other purpose.  How do you specify the width of the columns?  VisiData uses the actual display column width, which feels very natural.
